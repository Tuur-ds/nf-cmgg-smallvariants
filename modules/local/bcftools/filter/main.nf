process BCFTOOLS_FILTER {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bcftools:1.20--h8b25389_0':
        'biocontainers/bcftools:1.20--h8b25389_0' }"

    input:
    tuple val(meta), path(vcf), path(tbi)

    output:
    tuple val(meta), path("*.${extension}"), emit: vcf
    tuple val(meta), path("*.tbi")         , emit: tbi, optional: true
    tuple val(meta), path("*.csi")         , emit: csi, optional: true
    path  "versions.yml"                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def last_args = args3 ?: args2 ?: args

    extension = last_args.contains("--output-type b") || last_args.contains("-Ob") ? "bcf.gz" :
                    last_args.contains("--output-type u") || last_args.contains("-Ou") ? "bcf" :
                    last_args.contains("--output-type z") || last_args.contains("-Oz") ? "vcf.gz" :
                    last_args.contains("--output-type v") || last_args.contains("-Ov") ? "vcf" :
                    "vcf"

    if ("$vcf" == "${prefix}.${extension}") error "Input and output names are the same, set prefix in module configuration to disambiguate!"

    def filter_2 = args2 ? "| bcftools filter --threads ${task.cpus} ${args2}" : ""
    def filter_3 = args3 ? "| bcftools filter --threads ${task.cpus} ${args3}" : ""

    """
    bcftools filter \\
        --threads ${task.cpus} \\
        $args \\
        $vcf \\
        ${filter_2} \\
        ${filter_3} \\
        --output ${prefix}.${extension}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def last_args = args3 ?: args2 ?: args

    extension = last_args.contains("--output-type b") || last_args.contains("-Ob") ? "bcf.gz" :
                    last_args.contains("--output-type u") || last_args.contains("-Ou") ? "bcf" :
                    last_args.contains("--output-type z") || last_args.contains("-Oz") ? "vcf.gz" :
                    last_args.contains("--output-type v") || last_args.contains("-Ov") ? "vcf" :
                    "vcf"
    def index = last_args.contains("--write-index=tbi") || last_args.contains("-W=tbi") ? "tbi" :
                last_args.contains("--write-index=csi") || last_args.contains("-W=csi") ? "csi" :
                last_args.contains("--write-index") || last_args.contains("-W") ? "csi" :
                ""
    def create_cmd = extension.endsWith(".gz") ? "echo '' | gzip >" : "touch"
    def create_index = extension.endsWith(".gz") && index.matches("csi|tbi") ? "touch ${prefix}.${extension}.${index}" : ""

    if ("$vcf" == "${prefix}.${extension}") error "Input and output names are the same, set prefix in module configuration to disambiguate!"

    """
    ${create_cmd} ${prefix}.${extension}
    ${create_index}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """
}
