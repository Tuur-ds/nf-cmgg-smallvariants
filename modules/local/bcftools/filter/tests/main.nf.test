nextflow_process {

    name "Test Process BCFTOOLS_FILTER"
    script "../main.nf"
    process "BCFTOOLS_FILTER"

    tag "modules"
    tag "modules_local"
    tag "bcftools"
    tag "bcftools/filter"

    test("sarscov2 - 1 filter") {

        config "./one_filter.config"

        when {
            process {
                """
                input[0] = [
                    [id:"vcf_test"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true),
                    []
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("sarscov2 - 2 filters") {

        config "./two_filters.config"

        when {
            process {
                """
                input[0] = [
                    [id:"vcf_test"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true),
                    []
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("sarscov2 - 3 filters") {

        config "./three_filters.config"

        when {
            process {
                """
                input[0] = [
                    [id:"vcf_test"],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf', checkIfExists: true),
                    []
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}