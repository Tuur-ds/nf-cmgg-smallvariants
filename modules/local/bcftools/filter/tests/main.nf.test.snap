{
    "sarscov2 - 2 filters": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz:md5,8e722884ffb75155212a3fc053918766"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz.tbi:md5,1a1edb22d2a33f9673449827e1cd38e8"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,9a336d1ee26b527d7a2bdbeead155f64"
                ],
                "csi": [
                    
                ],
                "tbi": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz.tbi:md5,1a1edb22d2a33f9673449827e1cd38e8"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz:md5,8e722884ffb75155212a3fc053918766"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9a336d1ee26b527d7a2bdbeead155f64"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-24T14:15:19.958449133"
    },
    "sarscov2 - 3 filters": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz:md5,8e722884ffb75155212a3fc053918766"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz.tbi:md5,1a1edb22d2a33f9673449827e1cd38e8"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,9a336d1ee26b527d7a2bdbeead155f64"
                ],
                "csi": [
                    
                ],
                "tbi": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz.tbi:md5,1a1edb22d2a33f9673449827e1cd38e8"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz:md5,8e722884ffb75155212a3fc053918766"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9a336d1ee26b527d7a2bdbeead155f64"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-24T14:14:01.149643861"
    },
    "sarscov2 - 1 filter": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz:md5,8e722884ffb75155212a3fc053918766"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz.tbi:md5,1a1edb22d2a33f9673449827e1cd38e8"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,9a336d1ee26b527d7a2bdbeead155f64"
                ],
                "csi": [
                    
                ],
                "tbi": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz.tbi:md5,1a1edb22d2a33f9673449827e1cd38e8"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "vcf_test"
                        },
                        "vcf_test.vcf.gz:md5,8e722884ffb75155212a3fc053918766"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9a336d1ee26b527d7a2bdbeead155f64"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-24T14:15:05.44617156"
    }
}